sklearn/utils/_param_validation.py

```diff
@@ -110,13 +111,21 @@ def validate_params(parameter_constraints):
         A dictionary `param_name: list of constraints`. See the docstring of
         `validate_parameter_constraints` for a description of the accepted constraints.
 
+        Note that the *args and **kwargs parameters are not validated and must not be
+        present in the parameter_constraints dictionary.
+
     Returns
     -------
     decorated_function : function or method
         The decorated function.
     """
 
     def decorator(func):
+        # The dict of parameter constraints is set as an attribute of the function
+        # to make it possible to dynamically introspect the constraints for
+        # automatic testing.
+        setattr(func, "_skl_parameter_constraints", parameter_constraints)
```

Comment:
I think there's an issue with that approach. The decorator works with functions and methods. If the decorator returns an object I think it can't work as expected to replace a method. But maybe I just don't know how to do it. Here's what I came up with:
```python
def validate_params(parameter_constraints):

    def decorator(func):

        @functools.wraps(func, updated=())
        class wrapper:
            def __init__(self):
                self._skl_parameter_constraints = parameter_constraints

            def __call__(self, *args, **kwargs):
                func_sig = signature(func)

                # Map *args/**kwargs to the function signature
                params = func_sig.bind(*args, **kwargs)
                params.apply_defaults()

                # ignore self/cls and positional/keyword markers
                to_ignore = [
                    p.name
                    for p in func_sig.parameters.values()
                    if p.kind in (p.VAR_POSITIONAL, p.VAR_KEYWORD)
                ]
                to_ignore += ["self", "cls"]
                params = {k: v for k, v in params.arguments.items() if k not in to_ignore}

                validate_parameter_constraints(
                    parameter_constraints, params, caller_name=func.__qualname__
                )

                return func(*args, **kwargs)

        return wrapper()

    return decorator
```
maybe someone has an idea ?

Reply:
Also, not sure it's that less hackish that setting an attribute on a function :smile:

Reply:
My understanding is that replacing a method by an object that is callable doesn't make it a method but makes it an attribute of the class that is callable. Thus, when calling it, `self` is not passed as first argument.

Reply:
Discussing IRL with @jeremiedbb, it seems that making it callable makes things more complex. I would be inclined to keep setting the attribute on the function.

Reply:
I'm okay with leaving this as is. I think to actually use a "class callable" with a decorator, it would end up something like `available_if` and using class descriptor's:

https://github.com/scikit-learn/scikit-learn/blob/2f787f4c99401120184649cdc5c3e098f9823746/sklearn/utils/metaestimators.py#L143

On that note, can you check to make sure that the current implementation does not run into issues like https://github.com/scikit-learn/scikit-learn/issues/21344 which was fixed in https://github.com/scikit-learn/scikit-learn/pull/23077?

Reply:
Thanks for the hint, I forgot about available_if ! 
Let me try to implement this and we'll chose the best solution

Reply:
I also think the attribute solution is the easiest even if it populates the callables' namespaces.